### DISCUSSION TITLE
Why is there still a prevalent usage of `cjs` instead of the newer `mjs` to build?

### DISCUSSION THREAD

**Discussion by fwqaaq** [web-infra-dev/rspack#2281] [Q&A]
* I don't understand why there is such a prevalent usage of `cjs` instead of `mjs` in both scripts and packages.
* And, I believe that it's entirely possible to use `mjs`, and it's a more modern approach.

> Who can tell me why?

**Comment by h-a-n-a** [comment]
We have a WIP PR to add this feature. See: https://github.com/web-infra-dev/rspack/pull/2168

**Reply by fwqaaq** [reply]
Excuse, what I meant to inquire was why this project is using `cjs` instead of `mjs`, as opposed to the configuration files.

**Reply by h-a-n-a** [reply]
> Excuse, what I meant to inquire was why this project is using `cjs' instead of `mjs`, as opposed to the configuration files.

@fwqaaq This is a long and tedious story. To make this story simple, you may think of some frequently used APIs like `__dirname`, `__filename` and `require`. These APIs are not available in the ES Module execution context of node. So, it is said that only libraries which will be directly consumed by Node can be easily to run in that environment. 
However, Rspack is depended by a bunch of applications that are not rewritten in ESM, no to say ESModule format is contagious as you can only import a `mjs` formatted file with dynamic import syntax in CJS. See: https://nodejs.org/api/modules.html#the-mjs-extension
It is just a teeny-tiny sneak-peak of these two format and there are a lot of details here that I haven't mentioned, this makes it too hard for us and people who depend on our library to work perfectly.
Hope this resolves some of your questions.

**Reply by fwqaaq** [reply]
> you may think of some frequently used APIs like `__dirname`,` __filename` and require

Oh, maybe I got it, but I guess the above issues are not the main problems, rather they are related to some third-party dependencies, such as `webpack`, or some others?

**Reply by h-a-n-a** [reply]
> > you may think of some frequently used APIs like `__dirname`,` __filename` and require
> 
> Oh, maybe I got it, but I guess the above issues are not the main problems, rather they are related to some third-party dependencies, such as `webpack`, or some others?

Not exactly, ESM can directly import CJS modules, but if people using for example `@rspack/core` for their project. They have to use dynamic import to load it, which breaks the sync require semantic they might be relying on.

**Reply by fwqaaq** [reply]
I don't understand what you mean. Can `respack/@core` be designed using `esm` instead of `cjs`? If `esm` can be used in the `rspack`, the developers are enabled to use `esm` to import it. What I don't understand is why `rspack` uses `cjs`. It's not just because of `__dirname`, `require` etc., because they can be replaced with other `esm` methods. I haven't looked closely at `rspack`, I'm just curious why it's not developed using `esm` like `vite`.

**Reply by h-a-n-a** [reply]
Vite leverages conditional exports to make the package be able to work with both ESM and CJS. What's more, only a subset(async) of functions can be used for work in both formats. See: https://github.com/vitejs/vite/blob/main/packages/vite/index.cjs#L9

**Reply by h-a-n-a** [reply]
We will follow closely with webpack on adopting ESM.

**Reply by fwqaaq** [reply]
Thank you for your answer.
